File: backend/routers/recommendations.py
"""Recommendations router — agent-generated financial recommendations.

Endpoints:
  POST   /api/recommendations/generate     — Generate a new recommendation
  GET    /api/recommendations/              — List user's recommendations
  GET    /api/recommendations/{id}          — Get single recommendation
  POST   /api/recommendations/{id}/vote     — Vote on a recommendation
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query, status

from auth.dependencies import get_current_user
from agents import InvestmentAgent
from database import get_db
from models.user import User
from services.recommendation_engine import (
    generate_recommendation,
    list_recommendations,
    record_vote,
)
from services.structured_output import (
    GenerateRequest,
    GenerateResponse,
    VoteRequest,
    VoteResponse,
)

router = APIRouter(prefix="/api/recommendations", tags=["recommendations"])


@router.get("/health")
def health() -> dict:
    """Health check for the Investment Agent and Ollama connectivity (no auth)."""
    try:
        agent = InvestmentAgent()
        result = agent.health_check()
        return {
            "ok": result.get("ok", False),
            "model": result.get("model", ""),
            "available_models": result.get("available_models", []),
        }
    except Exception as e:
        return {"ok": False, "error": str(e), "model": "", "available_models": []}


@router.post("/generate", response_model=GenerateResponse, status_code=status.HTTP_201_CREATED)
def generate(
    request: GenerateRequest,
    user: User = Depends(get_current_user),
    db=Depends(get_db),
) -> GenerateResponse:
    """Generate a recommendation using the specified agent type.

    Runs the full pipeline: validate → context → agent → confidence → persist.
    """
    try:
        return generate_recommendation(db, str(user.id), request)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Agent generation failed: {e}",
        )


@router.get("/")
def list_all(
    agent_type: str | None = Query(None, description="Filter by agent type"),
    status_filter: str | None = Query(None, alias="status", description="Filter by status"),
    limit: int = Query(20, ge=1, le=100),
    offset: int = Query(0, ge=0),
    user: User = Depends(get_current_user),
    db=Depends(get_db),
) -> list[dict]:
    """List all recommendations for the current user."""
    try:
        return list_recommendations(
            db, str(user.id), agent_type=agent_type, status=status_filter, limit=limit, offset=offset
        )
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.get("/{recommendation_id}")
def get_one(
    recommendation_id: str,
    user: User = Depends(get_current_user),
    db=Depends(get_db),
) -> dict:
    """Get a single recommendation by ID."""
    from models.recommendation import Recommendation

    rec = (
        db.query(Recommendation)
        .filter(Recommendation.id == recommendation_id, Recommendation.user_id == str(user.id))
        .first()
    )
    if not rec:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Recommendation not found")

    return {
        "id": rec.id,
        "agent_type": rec.agent_type,
        "recommendation_type": rec.recommendation_type,
        "ticker": rec.ticker,
        "action": rec.action,
        "quantity": rec.quantity,
        "rationale": rec.rationale,
        "confidence_score": rec.confidence_score,
        "risks": rec.risks,
        "alternatives": rec.alternatives,
        "before_state": rec.before_state,
        "after_state": rec.after_state,
        "status": rec.status,
        "model_used": rec.model_used,
        "tokens_used": rec.tokens_used,
        "created_at": str(rec.created_at),
        "expires_at": str(rec.expires_at) if rec.expires_at else None,
    }


@router.post("/{recommendation_id}/vote", response_model=VoteResponse)
def vote(
    recommendation_id: str,
    vote_request: VoteRequest,
    user: User = Depends(get_current_user),
    db=Depends(get_db),
) -> VoteResponse:
    """Cast a vote (accepted/rejected/deferred) on a recommendation."""
    try:
        return record_vote(db, str(user.id), recommendation_id, vote_request)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))

File: frontend/src/api/recommendations.ts
/** Recommendation API — fetch, generate, vote on financial recommendations. */

import { api } from './client';

export interface GenerateRequest {
  agent_type: 'investment' | 'debt' | 'retirement';
  user_message?: string;
  skill?: string;
}

export interface StructuredOutput {
  title?: string;
  summary?: string;
  rationale?: string;
  action?: string;
  ticker?: string;
  quantity?: number;
  agent_type?: string;
  confidence?: {
    overall: number;
    math_certainty: number;
    market_assumptions: number;
    user_goal_alignment: number;
    execution_likelihood: number;
    explanation?: string;
  };
  risks?: string[];
  alternatives?: string[];
  before_state?: Record<string, unknown>;
  after_state?: Record<string, unknown>;
}

export interface GenerateResponse {
  recommendation_id: string;
  agent_type: string;
  structured: StructuredOutput;
  raw_text: string;
  tokens_used: number;
}

export interface VoteRequest {
  vote: 'accepted' | 'rejected' | 'deferred';
  comment?: string;
}

export interface VoteResponse {
  vote_id: string;
  recommendation_id: string;
  vote: string;
}

export interface Recommendation {
  id: string;
  agent_type: string;
  recommendation_type: string;
  ticker: string | null;
  action: string;
  quantity: number | null;
  rationale: string;
  confidence_score: number;
  risks: string;
  alternatives: string;
  before_state: string | null;
  after_state: string | null;
  status: string;
  model_used: string | null;
  tokens_used: number | null;
  created_at: string;
  expires_at: string | null;
}

function paramsToString(params?: Record<string, string | number | undefined>): string {
  if (!params) return '';
  const entries = Object.entries(params).filter(([, v]) => v !== undefined);
  if (entries.length === 0) return '';
  return '?' + new URLSearchParams(entries as [string, string][]).toString();
}

export const recommendationsApi = {
  generate: (request: GenerateRequest) =>
    api<GenerateResponse>('/recommendations/generate', {
      method: 'POST',
      body: JSON.stringify(request),
    }),

  list: (params?: { agent_type?: string; status?: string; limit?: number; offset?: number }) =>
    api<Recommendation[]>(`/recommendations/${paramsToString(params)}`),

  get: (id: string) => api<Recommendation>(`/recommendations/${id}`),

  vote: (id: string, vote: VoteRequest) =>
    api<VoteResponse>(`/recommendations/${id}/vote`, {
      method: 'POST',
      body: JSON.stringify(vote),
    }),

  health: () => api<{ ok: boolean; model: string; available_models: string[] }>('/recommendations/health'),

  // Phase 39 fix T2.2: probe endpoint — returns { empty: true } when there are no active recs.
  empty: () => api<{ empty: boolean }>('/recommendations/empty'),
};